src/main/scala/com/microsoft/hyperspace/index/rules/HPFilterIndexRule.scala

```diff
@@ -0,0 +1,166 @@
+      case Filter(_: Expression, ExtractRelation(relation))
+          if !RuleUtils.isIndexApplied(relation) =>
+        indexes.filterKeys(relation.plan.equals(_))
+      case _ =>
+        Map.empty
+    }
+  }
+}
+
+object HPFilterColumnFilter extends QueryPlanIndexFilter {
+  override def apply(plan: LogicalPlan, indexes: PlanToIndexesMap): PlanToIndexesMap = {
+    if (indexes.isEmpty) {
+      return Map.empty
+    }
+
+    val (filterColumnNames, projectColumnNames) = plan match {
+      case project @ Project(_, _ @Filter(condition: Expression, ExtractRelation(relation)))
+          if !RuleUtils.isIndexApplied(relation) =>
+        val projectColumnNames = CleanupAliases(project)
+          .asInstanceOf[Project]
+          .projectList
+          .map(_.references.map(_.asInstanceOf[AttributeReference].name))
+          .flatMap(_.toSeq)
+        val filterColumnNames = condition.references.map(_.name).toSeq
+        (filterColumnNames, projectColumnNames)
+
+      case Filter(condition: Expression, ExtractRelation(relation))
+          if !RuleUtils.isIndexApplied(relation) =>
+        val relationColumnNames = relation.plan.output.map(_.name)
+        val filterColumnNames = condition.references.map(_.name).toSeq
+        (filterColumnNames, relationColumnNames)
+      case _ =>
+        (Seq(), Seq())
+    }
+
+    // Filter candidate indexes if:
+    //  1. Filter predicate's columns include the first 'indexed' column of the index.
+    //  2. The index covers all columns from the filter predicate and output columns list.
+    val candidateIndexes =
+      indexes.head._2.filter { index =>
+        val ddColumns = index.derivedDataset.properties.columns
+        withReasonTag("The first indexed column should be in filter columns.") {
+          ResolverUtils.resolve(spark, ddColumns.indexed.head, filterColumnNames).isDefined
+        }(plan, index) &&
+        withReasonTag(
+          "Index does not contain required columns. Required columns: " +
+            s"[${filterColumnNames ++ projectColumnNames}], indexed & included columns: " +
+            s"[${ddColumns.indexed ++ ddColumns.included}") {
+          ResolverUtils
+            .resolve(
+              spark,
+              filterColumnNames ++ projectColumnNames,
+              ddColumns.indexed ++ ddColumns.included)
+            .isDefined
+        }(plan, index)
+      }
+
+    Map(indexes.head._1 -> candidateIndexes)
+  }
+}
+
+object HPFilterRankFilter extends IndexRankFilter {
+  override def apply(plan: LogicalPlan, indexes: PlanToIndexesMap): PlanToSelectedIndexMap = {
+    if (indexes.isEmpty || indexes.size != 1 || indexes.head._2.isEmpty) {
+      Map.empty
+    } else {
+      val selected = FilterIndexRanker.rank(spark, plan, indexes.head._2).get
+      setRankReasonTag(indexes.head._1, indexes.head._2, selected)
+      Map(indexes.head._1 -> selected)
+    }
+  }
+}
+
+/**
+ * HPFilterIndexRule looks for opportunities in a logical plan to replace
```

Comment:
Do we have any plan to support covering indexes that are not hash partitioned? I don't see any hash partitioning specific code here or in RuleUtils, but covering index related code. It seems the code don't care about whether the index data is hash partitioned or range partitioned (if such a thing exists), as long as it is laid out efficiently (clustered) regarding the key columns. On the other hand, it won't work for non-covering indexes, as just replacing the source data with index data only works for covering indexes. Therefore naming this Hash Partitioned but not Covering seems misleading.

Actually, I prefer just `FilterIndexRule` because it is the only rule for filters at the moment, and we don't know yet how we will support more types in the future. My second preference would be `CoveringFilterIndexRule`, if you think the index type should be in the name. Lastly, `HashPartitionedCoveringFilterIndexRule` if that information about partitioning is really that important (to me it seems not). Also, I'm against using uncommon acronyms such as HP for naming things.

Reply:
For other index type, we need to add another rule as it might have a different filter condition for the index & different application / score function. We can add index type filter at first.  

> Actually, I prefer just FilterIndexRule because it is the only rule for filters at the moment, and we don't know yet how we will support more types in the future.

We will add it soon and your data-skipping index also requires a new rule as it has different application & score functions.
But as we discussed, it would be better to rename CoveringIndex related things in anoter change. I'll change the name to FilterIndexRule_disabled

Reply:
> For other index type, we need to add another rule as it might have a different filter condition for the index & different application / score function. We can add index type filter at first.

Another possible design is abstracting index functionality into an interface and making the rule work on the interface. I think both designs have their pros and cons.

By the way, now I understand why you named it differently. I think if names collide then it is better to put new things in a new package and rename/move later at once when new things replace old things. It might confuse developers if old and new things, seemingly doing the same thing, are in the same place, although this is only temporary. But I'm okay with either way.

Reply:
Tried it but reverted since I needed to import all things in `rules` package.